cpp/src/arrow/compute/expression.cc

```diff
@@ -1242,33 +1242,104 @@ struct Inequality {
                             /*insert_implicit_casts=*/false, &exec_context);
   }
 
+  /// Simplify an is_in predicate against this inequality as a guarantee.
+  Result<Expression> SimplifyIsIn(Expression expr) {
+    const auto& guarantee = *this;
+    auto call = expr.call();
+    auto options = checked_pointer_cast<SetLookupOptions>(call->options);
+
+    auto value_set = options->value_set.make_array();
+    if (!value_set) return expr;
+    if (value_set->length() == 0) return literal(false);
```

Comment:
Or you can guard this with the `guarantee.nullable` check and write a comment with a short proof of correctness:

> Since `!guarantee.nullable`, we don't have to consider all possible null-matching behaviors in the simplification and can guarantee that `false` is the return value when the `value_set` is empty.

Reply:
I changed it so that everything is guarded by the `guarantee.nullable` check.